wgengine/magicsock/magicsock.go

```diff
@@ -2097,6 +2103,32 @@ func (c *Conn) SetNetworkUp(up bool) {
 	}
 }
 
+// SetPreferredPort sets the connection's preferred local port.
+func (c *Conn) SetPreferredPort(port uint16) {
+	if uint16(c.port.Get()) == port {
+		return
+	}
+	c.port.Set(uint32(port))
+
+	if err := c.bindSocket(&c.pconn4, "udp4", true); err != nil {
```

Comment:
why the pconn4 before the the portMapper.SetLocalPort and then the pconn6? something subtle happening?

Reply:
This code was matching Rebind. I factored it out to make that clear.

I believe the idea is: we treat rebinding IPv4 as vital, and rebinding IPv6 as nice-to-have. So as soon as bindSocket works for udp4, we change the local conn to let STUN do its thing, then spend some time trying to get udp6 to optionally work.